File: Assignments/Programming-Assignment-1/Problem4.py
#!/usr/bin/env python3

# Method 1: O(N log N)
def GetMajorityElement(arr):
    """
    TODO: Fill in docstring
    """
    if (len(arr) == 1):
        return arr[0];    # possible issues
    k = int(len(arr)/ 2) 
    left_element= GetMajorityElement(arr[:k])
    right_element = GetMajorityElement(arr[k:])
    if left_element == right_element:
        return left_element
    # These need to search the entirety of the array
    left_count = GetFrequency(arr, left_element)
    right_count = GetFrequency(arr, right_element)
    if left_count > (k + 1):
        return left_element
    elif right_count > (k + 1):
        return right_element
    else:
        return -1    # come up with better return than this

def GetFrequency(arr, key):
    """
    TODO: Fill in docstring
    """
    count = 0
    for i in range(len(arr)):
        if arr[i] == key:
            count += 1
    return count

def main():
    arr = [5,5,5,2,5,2,5,5]
    majority = GetMajorityElement(arr)
    print(f"{majority}")
    #  if (majority == None):
    #      print("No majority element")
    #  else:
    #      print(f"Majority element: {majority}")
    
main()

File: Assignments/Programming-Assignment-1/Problem4.cc
#include <iostream>
#include <cmath>
#include <vector>

int countFrequency(int X[], int l, int r, int value)
{
    int count = 0;
    for (int i = l; i <= r; i = i + 1)
    {
        if (X[i] == value)
            count = count + 1;
    }
    return count;
}

int findMajorityElement(int X[], int l, int r)
{
    if (l == r)
        return X[l];

    /* DIVIDE */
    int mid = l + (r - l) / 2;
    int leftMajority = findMajorityElement(X, l, mid);
    int rightMajority = findMajorityElement(X, mid + 1, r);

    /* CONQUER */
    if (leftMajority == rightMajority)
        return leftMajority;

    int leftCount = countFrequency(X, l, r, leftMajority);
    int rightCount = countFrequency(X, l, r, rightMajority);

    /* COMBINE */
    if (leftCount > (r - l + 1) / 2)
        return leftMajority;
    else if (rightCount > (r - l + 1) / 2)
        return rightMajority;
    else
        return -1; // No majority element in this subarray
}

int majorityElement(int X[], int n)
{
    return findMajorityElement(X, 0, n - 1);
}

int main()
{
    int arr[]{5,5,5,2};
    int majority = majorityElement(arr, 4);
    std::cout << "Result: " << majority << "\n";
}